### PR TITLE
Parallelize site rendering using rayon

### PR DESCRIPTION
**Pull request by williamyaoh**
See #70.

Parallelizes the rendering of individual pages/sections, using a `rayon` thread pool.

## Issues

This solution is as simple as possible, by passing every page and section into the work pool for `rayon`. For sections with very few pages in them, the overhead from the parallelization itself will outweigh the benefits from parallel execution. It might be worth it to fallback to sequential rendering in cases where the number of pages is low; the breakpoint would have to be determined through experimentation, and would probably differ based on number of cores on the testing machine. It would also further complicate the code.

### CODE REVIEW

File: src/rendering/mod.rs [hunk 1]
```diff
@@ -1,4 +1,5 @@
 pub mod highlighting;
+pub mod parsing;
```
Comment (Keats) [inline src/rendering/mod.rs]: that should go in the highlighting module imo

**Review by Keats — CHANGES_REQUESTED** [review]
Looks really good, thanks a lot.
I want to get the benches done before merging that to be able to compare the improvements.

Rayon is quite magical

**Comment by Keats** [comment]
Can you remove my commit from the PR? I'll merge it afterwards

**Comment by williamyaoh** [comment]
Done. Rebasing reminded me to add a `CHANGELOG` entry anyways, which I always seem to forget about 😛

**Comment by Keats** [comment]
thanks a lot!

**Comment by Keats** [comment]
Some pretty good win overall (now.bench is the `next` branch with rayon):
```sh
 name                                                current.bench ns/iter  now.bench ns/iter  diff ns/iter   diff % 
 bench_loading_big_blog                              181,721,374            167,167,614         -14,553,760   -8.01% 
 bench_loading_big_blog_with_syntax_highlighting     945,746,044            603,589,604        -342,156,440  -36.18% 
 bench_loading_medium_blog                           43,167,865             46,333,862            3,165,997    7.33% 
 bench_loading_medium_blog_with_syntax_highlighting  234,259,042            143,669,346         -90,589,696  -38.67% 
 bench_loading_medium_kb                             179,065,917            170,680,579          -8,385,338   -4.68% 
 bench_loading_medium_kb_with_syntax_highlighting    936,848,284            553,444,752        -383,403,532  -40.92% 
 bench_loading_small_blog                            4,456,486              4,211,842              -244,644   -5.49% 
 bench_loading_small_blog_with_syntax_highlighting   26,783,438             16,781,131          -10,002,307  -37.35% 
 bench_loading_small_kb                              16,084,277             14,526,129           -1,558,148   -9.69% 
 bench_loading_small_kb_with_syntax_highlighting     90,788,320             55,210,929          -35,577,391  -39.19% 
```

Notably, the `bench_loading_big_blog_with_syntax_highlighting` shows that it's loading 1000 pages and rendering the content with syntax highlighting in 600ms on a 4y old laptop.

I didn't run the `huge_*` in the same batch because I was running out of memory but:

```sh
test bench_loading_huge_blog                            ... bench: 1,746,137,505 ns/iter (+/- 163,960,028)
test bench_loading_huge_blog_with_syntax_highlighting   ... bench: 6,142,035,383 ns/iter (+/- 1,248,391,412)
```
So that's 6 seconds for reading and rendering + syntax highlighting of 10k files. Not bad considering I didn't care about performance at all so far!

**Pull request merged**